fiat-roles/src/main/java/com/netflix/spinnaker/fiat/roles/UserRolesSyncer.java

```diff
@@ -62,57 +76,78 @@ public void sync() {
   }
 
   public long syncAndReturn() {
-    updateServiceAccounts();
-    return updateUserPermissions();
-  }
-
-  private void updateServiceAccounts() {
-    AtomicInteger updateCount = new AtomicInteger(0);
-
     FixedBackOff backoff = new FixedBackOff();
     backoff.setInterval(retryIntervalMs);
     BackOffExecution backOffExec = backoff.start();
 
+    if (!isServerHealthy()) {
+      log.warn("Server is currently UNHEALTHY. User permission role synchronization and " +
+                   "resolution may not complete until this server becomes healthy again.");
+    }
+
     while (true) {
       try {
-        serviceAccountProvider
-            .getAll()
-            .forEach(serviceAccount -> permissionsResolver
-                .resolve(serviceAccount.getName())
-                .ifPresent(permission -> {
-                  permissionsRepository.put(permission);
-                  updateCount.incrementAndGet();
-                })
-            );
-        log.info("Synced " + updateCount.get() + " service accounts");
-        return;
-      } catch (ProviderException pe) {
+        Map<String, UserPermission> combo = new HashMap<>();
+        Map<String, UserPermission> temp;
+        if (!(temp = getUserPermissions()).isEmpty()) {
+          combo.putAll(temp);
+        }
+        if (!(temp = getServiceAccountsAsMap()).isEmpty()) {
+          combo.putAll(temp);
+        }
+
+        return updateUserPermissions(combo);
+      } catch (ProviderException|PermissionResolutionException ex) {
+        Status status = healthIndicator.health().getStatus();
         long waitTime = backOffExec.nextBackOff();
         if (waitTime == BackOffExecution.STOP) {
-          return;
+          log.error("Unable to resolve service account permissions.", ex);
+          return 0;
```

Comment:
You can successfully update zero permissions, correct? I would change this to be more indicative of failure - maybe here is a good place for an `Optional` return type?